lua/telescope/_extensions/file_browser/git.lua

```diff
@@ -0,0 +1,55 @@
+local Path = require "plenary.path"
+
+local os_sep = Path.path.sep
+
+local M = {}
+
+-- icon defaults are taken from Telescope git_status icons
+local icon_defaults = {
+  added = "+",
+  changed = "~",
+  copied = ">",
+  deleted = "-",
+  renamed = "➡",
+  unmerged = "‡",
+  untracked = "?",
+}
+
+--- Returns a display item for use in a display array based on a git status
+---@param opts table: configuration options to override defaults
+---@param status string: the string to convert to a display item
+---@return table: a display item
+M.make_display = function(opts, status)
+  local icons = vim.tbl_extend("keep", opts.git_icons or {}, icon_defaults)
+
+  -- this is copied from the Telescope git_status mapping and highlight groups
+  local git_abbrev = {
+    [" A"] = { icon = icons.added, hl = "TelescopeResultsDiffAdd" },
+    [" U"] = { icon = icons.unmerged, hl = "TelescopeResultsDiffAdd" },
+    [" M"] = { icon = icons.changed, hl = "TelescopeResultsDiffChange" },
+    [" C"] = { icon = icons.copied, hl = "TelescopeResultsDiffChange" },
+    [" R"] = { icon = icons.renamed, hl = "TelescopeResultsDiffChange" },
+    [" D"] = { icon = icons.deleted, hl = "TelescopeResultsDiffDelete" },
+    ["??"] = { icon = icons.untracked, hl = "TelescopeResultsDiffUntracked" },
+  }
+  local status_config = git_abbrev[status] or {}
+
+  local empty_space = " "
+  return { status_config.icon or empty_space, status_config.hl }
+end
+
+--- Returns a map of absolute file path to file status
+---@param output table: lines of the git status output
+---@param cwd string: the path from which the command was triggered
+---@return table: map from absolute file paths to files status
+M.parse_status_output = function(output, cwd)
+  local parsed = {}
+  for _, value in ipairs(output) do
+    local status = string.sub(value, 1, 2)
+    local file = cwd .. os_sep .. string.sub(value, 4, -1)
+    parsed[file] = status
```

Comment:
> I think we should trim the status here as well as in the git_abbrev table above.

I would like to avoid that to be able to show different icons and highlight groups depending wether the change is staged or not. I feel like this is a shortcoming in most git status features in file browsers and tools out there. Obviously, it makes it simpler to handle the different cases but it also limits the usability of it.

> Another edge case I just came across is when a file has both staged and unstaged hunks.

Ah nice! I haven't thought of that. Now I also understand what `[AMD]`, `[ MTD]` and `[MTARC]` mean in the git status docs. Those are all the possible statuses this status can be combined with. I would suggest that we then parse `X` and `Y` separately and reserve 2 characters for the status, one for the staged changes (X) and one for the unstaged changes (Y). This way also the table gets smaller and parsing becomes simpler. It would mean that we get annotations like 
```
"~~" - changes both staged and upstaged
"+~" - file added and has unstaged changes
"➡~" - file renamed and has unstaged changes
```

The first column I would also suggest to always color wit `TelescopeResultsDiffAdd` and the second color with `TelescopeResultsDiffChange` or `TelescopeResultsDiffDelete` depending on change or deletion.

Reply:
This is how it would look like with the double status. I decided against the two hl groups because the plugin doesn't support two hl groups within a single item I think, and otherwise it would add an extra gap to it making the git status even wider.
<img width="1645" alt="Screenshot 2023-01-22 at 13 41 26" src="https://user-images.githubusercontent.com/5249233/213916391-3688eed6-876e-40c0-8f36-a1f3e4f73368.png">
<img width="1645" alt="Screenshot 2023-01-22 at 13 41 31" src="https://user-images.githubusercontent.com/5249233/213916378-e7be1310-9acd-430b-a4be-4d8c8a178127.png">

Reply:
I'm getting different highlighting from you. Specifically the renamed/unstaged changes you show in the last screenshot.
![image](https://user-images.githubusercontent.com/66286082/213952440-0b739012-7fea-44ed-9732-ed09bf6d5ec9.png)

Regarding the staged vs unstaged double statuses and the highlighting limitations. I don't particularly love the idea of settling with one color for two statuses which may not share colors. My brain is getting mixed signals seeing different colors for the "changed" status for example depending on the staged/unstaged status (as per the my screenshot).
I think I would either show both statuses with their respective highlights, or go back to showing one status - in the priority of staged > unstaged > none. Leaning towards the latter since show both statuses individually adds an awkward looking gap.

Reply:
> I'm getting different highlighting from you. Specifically the renamed/unstaged changes you show in the last screenshot.

Between my screenshot and the state I pushed I switched from showing the unstaged changes hl group to the staged hl group.

Reply:
@jamestrew I reverted to prioritise unstaged changes. Does that make it more intuitive? I've been working with it for the past week and find it actually good.